### PR TITLE
Add add_mock_package

### PR DESCRIPTION
**Pull request by joemarshall**
Added a method to micropip to allow easy mocking of packages (for use in jupyterlite / pyodide)

It takes package name and version, along with a list of implemented modules and optionally their code if needed.

### CODE REVIEW

File: micropip/_micropip.py
Comment (rth) [inline micropip/_micropip.py:764]: Do we actually need to create these physical files?

I was thinking we could maybe directly initialize `sys.modules` with the mapping from `modules: dict[str, ModuleType | str]`? That way it's also easy to use `MagicMock` as a module or any other custom object defined at runtime to make a given dependency work.

 Though yes, the API to create modules dynamically via [`importlib.util.module_from_spec`](https://docs.python.org/3/library/importlib.html#importlib.util.module_from_spec)  or by [inheriting from ModuleType](https://stackoverflow.com/a/67813054/1791279) isn't very straightforward.

Comment (hoodmane) [reply micropip/_micropip.py:764]: Maybe making another custom Finder/Loader would be a good way to go. The API for that isn't too bad.

Comment (joemarshall) [reply micropip/_micropip.py:764]: I actually wrote it initially using mock objects and putting them into sys.modules, but I reverted to writing them out to files. I did it like that because it is really simple and reliable, things like file paths etc. just work. It also works nicely in terms of loading modules, in that init code is only run on actual import. Oh and in the case that you're running somewhere with persistent file system, your mock modules continue to exist on reload of python.

It's a pretty minor change - you just create a module and exec the code into it in the loader, I can put that back in if you reckon the possibility of using mock objects makes sense.

Comment (joemarshall) [reply micropip/_micropip.py:764]: Actually I think rather than take an object it needs to take a callable which takes a module object and does whatever needs to be done on it to make it. I.e. the equivalent of loader.exec_module

That way you can mock module initialisation correctly.

Comment (joemarshall) [reply micropip/_micropip.py:764]: One further thought on this - at some point I want to implement some kind of mega wheel option, to bundle a set of wheels and all deps into a single wheel with no deps. For that use case keeping persistent mock modules would be nice. If it's okay I'd like to keep that as an option in micropip?

It would also be nice for any point that pyodide is run in a persistent file system. 

I'm thinking that a flag "persistent" could be added (which only works for string modules) and then the behaviour with finder and loader could be used in the case that the flag isn't set.

**Comment by ryanking13** [comment]
> It would also be nice for any point that pyodide is run in a persistent file system.

I think this is reasonable, as having a physical file would be a most easy way to keep the state of installed packages or to bundle packages (for example, in `pyodide-pack`) for later use.

We can change the implementation of `add_mock_package` later if we find a better way to deal with this, but I'm +1 with this approach for now.

File: micropip/_micropip.py
Comment (ryanking13) [inline micropip/_micropip.py:733]: docstring for `version` parameter is omitted

File: micropip/_micropip.py
Comment (ryanking13) [inline micropip/_micropip.py:744]: Perhaps use `dist-info` directory instead?

File: micropip/_micropip.py
Comment (ryanking13) [inline micropip/_micropip.py:763]: Maybe we should raise error if a directory exists, because it would mean that the package is already installed somehow.

File: micropip/_micropip.py
Comment (ryanking13) [inline micropip/_micropip.py:700]: Could you please add type hints?

File: micropip/_micropip.py
Comment (ryanking13) [inline micropip/_micropip.py:766]: I think we should keep the list of mocked modules.

File: micropip/_micropip.py
Comment (joemarshall) [reply micropip/_micropip.py]: done

File: micropip/_micropip.py
Comment (joemarshall) [reply micropip/_micropip.py:805]: Added list_mock_modules instead - which uses module installer metadata to list modules it has installed

**Comment by joemarshall** [comment]
@ryanking13 @rth I've added support for non-persistent mocks, and made the tests work in pyodide. I think I've fixed the things from the reviews above.

There's a bit more code in there to do the in-memory mocking, which is probably worth a look over before merging.

**Comment by joemarshall** [comment]
Oh and it supports listing and removing mocks also btw.

File: micropip/_micropip.py
Comment (ryanking13) [inline micropip/_micropip.py:813]: Maybe move these hard coded INSTALLER types into a variables?

File: micropip/_mock_package.py
Comment (ryanking13) [inline micropip/_mock_package.py:79]: I'm not sure we want an importing micropip to have a side effect of modifying `sys.meta_path`, though it does not have any effect when no mock packages are registered.

Perhaps modify sys.meta_path when `add_in_memory_distribution` is first called?

File: tests/test_micropip.py
Comment (ryanking13) [inline tests/test_micropip.py:986]: There is a [`tmp_path`](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) fixture for this purpose.

File: tests/test_micropip.py
Comment (ryanking13) [inline tests/test_micropip.py:1106]: Maybe use `with pytest.raises(...)` instead?

File: tests/test_micropip.py
Comment (ryanking13) [inline tests/test_micropip.py:1058]: Same here (`pytest.raises`)

**Review by ryanking13 — APPROVED** [review]
Thanks, @joemarshall. Having a persistent parameter seems good to me. Overall LGTM, if @rth or @hoodmane is also +1 with this, I'll merge.

Could you please also update the [changelog](https://github.com/pyodide/micropip/blob/main/CHANGELOG.md)? Probably we can follow the format of [KeepAChangelog](https://keepachangelog.com/en/1.0.0/)

**Comment by joemarshall** [comment]
Okay, done those bits and pieces now and changelog. Good spot about the sys.meta_path thing, that was silly.

File: micropip/_micropip.py
Comment (rth) [inline micropip/_micropip.py:751]: I think there is an indentation issue in this docstring, but we can fix it once we start actually building docs for it.

**Review by rth — APPROVED** [review]
Thanks a lot for this work @joemarshall  and @ryanking13 for the review!

**Comment by ryanking13** [comment]
I guess we should release 0.1.1 (or 0.2.0) with this. Though probably we need to start building docs for micropip before making a release.

**Pull request merged**
